### ISSUE TITLE
help: "Jump to" autocompletion incorrect focus nuisance

### ISSUE THREAD

**Issue by Osndok** [jaap-karssenberg/zim-desktop-wiki#536]
About 20% of the time, when navigating to a new page using the 'jump to' feature, it misdirects me.

1. Ctrl-J (to open the 'jump to' popup)
2. type: ":Some:Pa"
3. arrow down to select ":Some:Path"
4. hit enter
5. zim immediately navigates to ":Some:Pa" (rather than simply filling out the field with the selected text)

I **think** what is going on is that the 'enter' keypress is bypassing the autocompletion menu entirely, and hitting the 'enter' key.

It **seems** to happen more frequently when I am typing quickly, and moving to a gray page stub.

It **might** be effected by my own doing b/c of https://github.com/Osndok/zim-desktop-wiki/commit/4efb1221a1514ba605525591b8806fdfa4881cf8 , but I recall experiencing this from time to time *before* placing this commit, I think it just made it more frequent.

Has anyone else experienced this on the mainline zim? or... could I get some help implementing my work-around idea of connecting the 'enter' action to explicitly looking for (and replacing the form text with) any autocompletion selection?

**Comment by nelas** [comment]
I was about to report this. What @Osndok described happens when the mouse cursor is floating over the drop down list. See this example:

![jump](https://user-images.githubusercontent.com/361179/51180720-6ac06e80-18c9-11e9-956a-d0930e116e15.gif)

If the mouse cursor is not floating over the list, using the keyboard `arrow keys` to select and pressing `enter` works as expected. If the mouse is used to select (click) an item in the list, it also works as expected. There must be some kind of conflict between keyboard and mouse input when the cursor is over the list, but the keyboard is used to select the page.

I've tested this on Linux Zim [0.70-rc1](874202386e40405404cc54bfaad9b5ceaa3aa5fa), but previous versions also had this issue (i.e. gtk3 did not solve it).

**Comment by Osndok** [comment]
Pheww... I'm glad that I'm not the only one experiencing this, @nelas...

Your theory (about the mouse position) was one of my first thoughts asto the cause, but when I was trying to be conscious of the mouse position, I found that it often is triggered even when the mouse is totally clear of the menu.

My best guess is that it is actually related to **focus**... so depending on one's window manager and settings (e.g. focus follows mouse, focus stealing prevention, etc), the popup menu might not have focus for the enter.

Come to think of it... if **anything** were to have a focus issue, it would probably be this... because it is expected that you can keep typing in the field **and** actuate the menu **and** use enter for two different-but-related purposes.

This is why I recommend explicitly checking for the popup's state during the other callback that is being fired for 'enter' in the dialog box; but this is just out of reach of my python skills for the moment. So if this continues too much longer, I might actually just avoid the other enter callback altogether (to avoid the misjump but require some alt/tab keywork), as this is currently my top issue with zim.

**Comment by jaap-karssenberg** [comment]
Interestingly you can find the same behavior when you run "gtk3-demo" and then the "entry completion" demo. So looks like you found a bug in Gtk itself. Would recommend posting a bug report directly for Gtk3 to have it fixed at the right level. 

Needs investigation what work around is possible in zim.

**Comment by nelas** [comment]
Ok, just reported the issue in the GTK repository: [Entry Completion fails when mouse pointer over dropdown](https://gitlab.gnome.org/GNOME/gtk/issues/1813)

**Comment by jaap-karssenberg** [comment]
@nelas Thanks for submitting the issue upstream, will keep ticket open here for reference

**Comment by nelas** [comment]
I'm no longer experiencing this issue with GNOME 3.36.1 on Ubuntu 20.04.1 (Wayland). It is also not occurring in the [`gtk3-demo` upstream](https://gitlab.gnome.org/GNOME/gtk/issues/1813), so I think it is solved. Can anybody test it out?

**Comment by nelas** [comment]
Just noticed that the issue persists on using Xorg instead of Wayland. Already notified upstream.

**Comment by introt** [comment]
Gtk 3.0 version 2.12 introduces the `inline_selection` property for `EntryCompletion`[^inline_selection], which makes the entry get filled with the selected item on keyboard navigation. This alone doesn't solve the issue: further changes are required to prevent the model (completion options) from getting updated on navigation.

I tried replacing this handler with ones for `insert-text` and `delete-text`[^signal], but even though they should only be emitted when text is inserted or deleted _by the user_(`changed` is emitted after a "user-visible operation[^signal]), that seems to also include `inline_selection`.

https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/009b8f423501c876902be7c3db7067fc2bba9adf/zim/gui/widgets.py#L1752

I also tried disconnecting the signal when receiving  a `match-selected` signal from the EntryCompletion[^match-selected], but navigation doesn't emit one!

A last resort would be patching the dialog's [do_response_ok](https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/009b8f423501c876902be7c3db7067fc2bba9adf/zim/gui/mainwindow.py#L1007) to check if `path` exists and if not, try finding one with `lookup_from_user_input`[^zim] or a variant of the model updater if those produce different results. Before trying that I'm going to let someone who knows anything about Gtk try another approach first.

---

Related issue: #1242

[^inline_selection]: https://lazka.github.io/pgi-docs/Gtk-3.0/classes/EntryCompletion.html#Gtk.EntryCompletion.props.inline_selection
[^signal]: https://lazka.github.io/pgi-docs/Gtk-3.0/classes/Editable.html#signals
[^match-selected]: https://lazka.github.io/pgi-docs/Gtk-3.0/classes/EntryCompletion.html#Gtk.EntryCompletion.signals.match_selected
[^entry]: https://lazka.github.io/pgi-docs/Gtk-3.0/classes/Entry.html#signals
[^pre]: https://lazka.github.io/pgi-docs/Gtk-3.0/classes/Entry.html#Gtk.Entry.signals.preedit_changed
[^zim]: https://introt.github.io/zim-desktop-wiki/docs/api/zim.notebook.index.pages.PagesView.html#lookup_from_user_input

**Comment by hsandt** [comment]
Confirmed on Ubuntu 22.04 with Unity desktop, but apparently only if you press Enter twice quickly...

An alternative (complementary to the proper fix) would be to have a proper Jump to... command that only jumps to an existing page (similar to Jump to navigation command in code/document editors) and add an option for the more rare cases where you actually want to create a new page (or simply a dedicated command to create a new page, as in code/document editors).

EDIT: aha! I see that @osky85 suggested something very similar in #1242 last year
We can probably fix both issues by doing both, as in most code editors:
- Jump to: enter string > show list of matches, most relevant on top
- Also select the first one (most relevant) automatically
- You cannot select nothing unless there are no matches. Even then, if you press enter, nothing happens, so you cannot accidentally create a new page

**Comment by hsandt** [comment]
Repro is inconsistent when it happens the first time (apparently moving selection and pressing Enter too fast causes the issue), but once the bug occurred and you're still focusing the new temporary page (with name in italics), if you repeat the Jump > Enter page name, even if you select the existing page, wait for a moment and press Enter, you will 100% fall on the same temporary page again.

Only after switching window to write this message, did Jump to existing page actually work...

**Comment by hsandt** [comment]
So I found a workaround, not fixing the core issue but getting around it by preventing creation of a new page.

**TL;DR**

Patch in two steps:

1. Do not allow non-existing page in text field by making such an input invalid:

```py
# mainwindow.py
OpenPageDialog
	__init__
		# after self.add_form() statement, add this
		self.form.widgets['page'].existing_only = True
```

2. Allow auto-completion when text input is invalid:

```py
# widgets.py
class PageEntry(InputEntry):

	def update_completion(self):
		# after model.clear()

		# instead of
		# if not text or not self.get_input_valid():
		# change the condition to:
		if not text:
			return
```

**Research**

I've been debugging this for a while, and found the following:
- the unexpected new page creation bug may occur when navigating the page autocomplete candidates with the keyboard arrows and confirming one with Enter (not when clicking on it with the mouse). Sometimes, Enter confirms the popup Jump action (the Dialog "OK" action) instead of confirming the autocompletion first. This causes `path` to be the manual user input text ":Some:Pa" instead of the full path ":Some:Path".
- to fix this I'd need to set the input context to ignore the Dialog buttons while the autocompletion dropdown is displayed
- ... however, the autocompletion system is not part of Zim, rather GTK (https://docs.gtk.org/gtk4/class.EntryCompletion.html) and the Gtk.EntryCompletion class is deprecated since 4.10 so I don't expect a patch to fix the current input context issues (unless we are far behind and a simple update to 4.9 would fix the issue, but there is no guarantee - I don't know which version of Gtk we use right now, but in the CHANGELOG I see that the prerequisites are 3.18 so that would be well behind).
- A user had a similar issue: https://stackoverflow.com/questions/49755847/is-there-a-way-to-caught-signal-when-gtkentrycompletion-popped-up but there is no actual reply (we should use signals but there is not a good one to use)

So I considered to solutions we can apply at Zim code level:

A. Manually detect when autocomplete dropdown is shown and prevent confirming Jump if so

I assumed model matches view and checked if the PageEntry completion model has any child:

```py
# mainwindow.py
class OpenPageDialog(Dialog):

	def do_response_ok(self):
		page_entry = self.form.widgets['page']
		page_entry_completion = page_entry.get_completion()
		page_entry_completion_model = page_entry_completion.get_model()
		if page_entry_completion_model.iter_n_children() > 0:
			return False
```

=> FAILED because model does *not* match view: the model preserves entries behind the hood while the autocomplete dropdown is closed (after moving caret, and most importantly after pressing Enter to confirm an autocomplete choice)

B. Disable new page creation entirely

This is actually the brutal version of https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/2460 as it hardcodes disabling new page creation - but a first step to adding an option that disables it, if some people really want to restore the old behavior.

At first, I tried to hack `open_page` itself to return early `if not page.exists():`. The problem is that it was too late to process an error and pressing Enter would *still* close the Jump popup instead of keeping it open.

So instead I decided to reuse the existing input valid check system (the one that shows bad input on a red background.

Fortunately, PageEntry already has a field named `existing_only` to pass in the constructor which will reject non-existing pages. Since `add_form` method doesn't allow passing extra constructor parameters, I set it manually as in exportdialog.py > InputPage.__init__

```py
# mainwindow.py
OpenPageDialog
	__init__
		# after self.add_form() statement, add this
		self.form.widgets['page'].existing_only = True
```

However, it did not only prevent Jumping to a non-existing page, it also stopped autocompletion whenever the entered text didn't exactly match a page (making autocompletion pointless). This is because of a condition preventing autocompletion altogether if the current path is incorrect. 

So I changed this condition:

```py
# widgets.py
class PageEntry(InputEntry):

	def update_completion(self):
		# after model.clear()

		# instead of
		# if not text or not self.get_input_valid():
		# change the condition to:
		if not text:
			return
```

I didn't see any bad side effect by allowing auto-completion on bad strings (e.g. starting with `+` in certain conditions).

**Limitations**

In theory, this solution B has a limitation compared to solution A: it still doesn't fix the fundamental issue of trying to Jump to the page path entered by the user, if the page *does* exist. So if user enters ":Some:Page" and still navigates to ":Some:Page2" in autocomplete dropdown, on press Enter, Zim *could still* give priority to Jump and jump to ":Some:Page" instead of ":Some:Page2".

In practice, I tried it many times and, oddly enough, I could not reproduce the bug with an existing page name. Apparently what I changed had a side effect or not allowing Jump without selecting autocomplete string first.
**UPDATE**: I could reproduce the bug, see next post

Of course a clean fix to handle input context better would be best, but since GTK deprecated the class we may need to upgrade to GTK 5 and use the new classes instead (but the doc is not clear on how we should replace EntryCompletion).

**PR**

I'm cleaning things up to prepare a PR, and checking if I can add an option to re-enable create new page on jump (either in preferences or as a checkbox in the Jump popup) to fully complete #2460. It may take some time so in the meantime I'm giving you the tips to do it yourself and I'll probably also open a branch with the minimalistic fix for other people to use.

**Comment by hsandt** [comment]
TL;DR: patch is incomplete as I feared, you still need to keep the mouse away from the area where the autocompletion dropdown list will appear

Oops. somewhat I totally forgot about the previous posts that had done a lot of research on exact repro and GTK (that's why my explanations on GTK are a bit redundant).

Using the fact that the issue occurs when the mouse cursor is hovering the autocompletion list dropdown, I could reproduce the bug even with my patch: in this case, the dropdown is closed and Zim tries to confirm the Jump, however in most cases, the input path is invalid (because user entered partial words) so nothing happens and the popup stays open, which mitigates the effects of the bug, giving a second chance to the user to jump. But it is still annoying because you have to type some character to trigger the completion popup once more and select the page once more (this time with the mouse cursor away) to properly jump.

In more rare cases where the input path actually matched an existing page, the bug will be clearly visible as you Jump To the page designated by the input path, and I have no fix for this at the moment.

While unlikely when typing partial words, but could totally happen if you type "Pa" > autocomplete "Page" > add ":" to display all sub-pages > select sub-page with mouse over dropdown list => "Page:" is still recognised as "Page" and therefore on Enter, you Jump To "Page" instead of the selected sub-page.

So anyway I should keep the mouse cursor away even with my patch. That said the patch reduces my stress a lot as I know that even when I fail to Jump, in most cases I have a second chance to do it right.

I'm gonna update the GTK issue with the extra piece of information that Enter tries to apply the standard confirm callback of the popup itself, not the dropdown, in these conditions, in case that helps them.

**Comment by hsandt** [comment]
I made a [fork](https://github.com/hsandt/zim-desktop-wiki) with a [feature branch jump-cannot-create-new-page](https://github.com/hsandt/zim-desktop-wiki/tree/feature/jump-cannot-create-new-page) with a [commit](https://github.com/hsandt/zim-desktop-wiki/commit/b73b1facf6d33ab9c7574a15592fb47f70fd084e) to apply the fix I mentioned above.
